// rtp.js/lib/tests/packets/RTCP/PliPacket.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const PliPacket_1 = require("../../../packets/RTCP/PliPacket");
const FeedbackPacket_1 = require("../../../packets/RTCP/FeedbackPacket");
const RtcpPacket_1 = require("../../../packets/RTCP/RtcpPacket");
const helpers_1 = require("../../../utils/helpers");
const pliPacketDump = {
    byteLength: 12,
    padding: 0,
    packetType: RtcpPacket_1.RtcpPacketType.PSFB,
    count: 1,
    messageType: FeedbackPacket_1.PsFeedbackMessageType.PLI,
    senderSsrc: 0x00003344,
    mediaSsrc: 0x55667788
};
const array = new Uint8Array([
    0x81, 0xce, 0x00, 0x02,
    0x00, 0x00, 0x33, 0x44,
    0x55, 0x66, 0x77, 0x88 // Media SSRC: 0x55667788
]);
const view = new DataView(array.buffer, array.byteOffset, array.byteLength);
describe('parse RTCP PLI packet', () => {
    test('buffer view is RTCP', () => {
        expect((0, RtcpPacket_1.isRtcp)(view)).toBe(true);
    });
    test('packet processing succeeds', () => {
        const packet = new PliPacket_1.PliPacket(view);
        expect(packet.needsSerialization()).toBe(false);
        expect(packet.dump()).toEqual(pliPacketDump);
        expect((0, helpers_1.areDataViewsEqual)(packet.getView(), view)).toBe(true);
        packet.serialize();
        expect(packet.needsSerialization()).toBe(false);
        expect(packet.dump()).toEqual(pliPacketDump);
        expect((0, helpers_1.areDataViewsEqual)(packet.getView(), view)).toBe(true);
        const clonedPacket = packet.clone();
        expect(clonedPacket.needsSerialization()).toBe(false);
        expect(clonedPacket.dump()).toEqual(pliPacketDump);
        expect((0, helpers_1.areDataViewsEqual)(clonedPacket.getView(), view)).toBe(true);
    });
});
describe('create RTCP PLI packet', () => {
    const packet = new PliPacket_1.PliPacket();
    test('packet view is RTCP', () => {
        expect((0, RtcpPacket_1.isRtcp)(packet.getView())).toBe(true);
    });
    test('packet processing succeeds', () => {
        // No optional fields in this packet so serialization is never needed.
        packet.setSenderSsrc(pliPacketDump.senderSsrc);
        packet.setMediaSsrc(pliPacketDump.mediaSsrc);
        expect(packet.needsSerialization()).toBe(false);
        expect(packet.dump()).toEqual(pliPacketDump);
        expect((0, helpers_1.areDataViewsEqual)(packet.getView(), view)).toBe(true);
        packet.serialize();
        expect(packet.needsSerialization()).toBe(false);
        expect(packet.dump()).toEqual(pliPacketDump);
        expect((0, helpers_1.areDataViewsEqual)(packet.getView(), view)).toBe(true);
        const clonedPacket = packet.clone();
        expect(clonedPacket.needsSerialization()).toBe(false);
        expect(clonedPacket.dump()).toEqual(pliPacketDump);
        expect((0, helpers_1.areDataViewsEqual)(clonedPacket.getView(), view)).toBe(true);
    });
});
